### PR TITLE
SNOW-237995: Add force_arrow support to the C API library, and add a field to SF_STMT to extract response headers

### PR DESCRIPTION
**Pull request by sfc-gh-ema**
For Java/Scala Stored procs, we wish to extract the response headers from libsnowflakeclient queries so that the JDBC instance in the UDF server can interpret them. Also, force_arrow is enabled from XP so that the client driver can parse Arrow results directly.

https://snowflakecomputing.atlassian.net/browse/SNOW-237995

Note that there is also a companion PR, that's necessary to make this work:
https://github.com/snowflakedb/snowflake/pull/18521

### CODE REVIEW

File: lib/client.c [hunk 1]
```diff
@@ -1600,6 +1643,16 @@ int64 STDCALL snowflake_affected_rows(SF_STMT *sfstmt) {
...
+
+    // Since Arrow isn't yet supported, we throw an error
+    if (is_force_arrow_mode(sfstmt->connection)) {
+        SET_SNOWFLAKE_STMT_ERROR(&sfstmt->error, SF_STATUS_ERROR_ATTEMPT_TO_RETRIEVE_FORCE_ARROW,
+                                 "Affected row-count not available using force-arrow results type", "",
+                                 sfstmt->sfqid);
+
+        return ret;
+    }
+
```
Comment (sfc-gh-kwagner) [inline lib/client.c:1655]: +1

File: tests/test_arrow_force.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    // numRows should return -1
+    assert_int_equal(snowflake_num_rows(sfstmt), -1);
+
+    // But running another query should clear the status
+    status = snowflake_query(sfstmt, "select 1", 0);
+    if (status != SF_STATUS_SUCCESS) {
+        dump_error(&(sfstmt->error));
+    }
+    assert_int_equal(status, SF_STATUS_SUCCESS);
```
Comment (sfc-gh-kwagner) [inline tests/test_arrow_force.c:63]: Can you disable force_arrow and then run a query and get the results back to make sure that it works as expected?

File: lib/client.c (hunk 1 above)
Comment (sfc-gh-ema) [reply lib/client.c:1655]: What's +1? I assume you mean 'and this too,' in reference to some other comment. But I can't find the other comment. :)

Comment (sfc-gh-kwagner) [reply lib/client.c:1655]: Oh I mean I like this piece of code haha

File: tests/utils/test_setup.c [hunk 3]
```diff
@@ -38,6 +38,23 @@ SF_CONNECT *setup_snowflake_connection() {
             "UTC", SF_BOOLEAN_TRUE);
 }
 
+/**
+ * Enables C_API_QUERY_RESULT_FORMAT='ARROW_FORCE'.
+ * This can be removed when Arrow support is officially added.
+ */
+SF_STATUS STDCALL enable_arrow_force(SF_CONNECT *sf) {
```
Comment (sfc-gh-kwagner) [inline tests/utils/test_setup.c:45]: Here you define it with STDCALL and in the header file you do not. In Linux, STDCALL is a no-op, but its important in Windows. Try adding STDCALL to the header file too

File: lib/client.c (hunk 1 above)
Comment (sfc-gh-ema) [reply lib/client.c:1655]: :)

File: tests/utils/test_setup.c (hunk 3 above)
Comment (sfc-gh-ema) [reply tests/utils/test_setup.c:45]: Ah, I see that. Thanks!

File: lib/client.c [hunk 4]
```diff
@@ -52,6 +52,23 @@ _reset_connection_parameters(SF_CONNECT *sf, cJSON *parameters,
...
+ * Checks whether the client is running in force_arrow mode.
+ *
+ * @param connection pointer to SF_CONNECT
+ * @return SF_BOOLEAN_TRUE if the connection is running in force_arrow mode, otherwise SF_BOOLEAN_FALSE
+ */
+static sf_bool is_force_arrow_mode(const SF_CONNECT *connection) {
+    if (connection->query_result_format == NULL) {
+        return SF_BOOLEAN_FALSE;
+    }
+    if (strcasecmp(connection->query_result_format, "arrow_force") == 0) {
```
Comment (sfc-gh-kwagner) [inline lib/client.c:65]: `strcasecmp` does not exist in Windows. Either use `strcmp` (std library) or `sf_strncasecmp` (in platform.h) instead

File: ci/dev/start.sh [hunk 5]
```diff
@@ -9,21 +9,23 @@ THIS_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
...
-#rebuild the image
-[[ $(docker ps | grep "$CONTAINER_NAME") ]] && docker container stop "$CONTAINER_NAME"
-docker image build -t $IMAGE_NAME \
- --build-arg BASE_IMAGE_NAME=${BUILD_IMAGE_NAMES[@]} \
- --build-arg LOCAL_USER_ID=$(id -u $USER) .
+##rebuild the image
+#[[ $(docker ps | grep "$CONTAINER_NAME") ]] && docker container stop "$CONTAINER_NAME"
+#docker image build -t $IMAGE_NAME \
+# --build-arg BASE_IMAGE_NAME=${BUILD_IMAGE_NAMES[@]} \
+# --build-arg LOCAL_USER_ID=$(id -u $USER) .
```
Comment (sfc-gh-kwagner) [inline ci/dev/start.sh:28]: Did you mean to include these changes?

Comment (sfc-gh-ema) [reply ci/dev/start.sh]: Oops.

**Comment by sfc-gh-kwagner** [comment]
LGTM

**Review by sfc-gh-kwagner — APPROVED** [review]

**Pull request merged**